samples/subsys/usb/mass/src/main.c

```diff
@@ -1,119 +1,173 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
-/* Sample to put the device in USB mass storage mode backed on a 16k RAMDisk. */
-
 #include <zephyr.h>
 #include <logging/log.h>
 #include <usb/usb_device.h>
+#include <fs/fs.h>
 #include <stdio.h>
 
 LOG_MODULE_REGISTER(main);
 
 #if CONFIG_DISK_ACCESS_FLASH
 #include <storage/flash_map.h>
+#define DISK_STORAGE_DEV	((void *)FLASH_AREA_ID(storage))
+#else
+#define DISK_STORAGE_DEV	(NULL)
+#endif
+
 #if CONFIG_FAT_FILESYSTEM_ELM
-#include <fs/fs.h>
 #include <ff.h>
+#endif
 
-static FATFS fat_fs;
-#define FATFS_MNTP      "/NAND:"
-
-static struct fs_mount_t fs_mnt = {
-	.type = FS_FATFS,
-	.mnt_point = FATFS_MNTP,
-	.storage_dev = (void *)FLASH_AREA_ID(storage),
-	.fs_data = &fat_fs,
-};
-#elif CONFIG_FILE_SYSTEM_LITTLEFS
-#include <fs/fs.h>
+#if CONFIG_FILE_SYSTEM_LITTLEFS
 #include <fs/littlefs.h>
-
 FS_LITTLEFS_DECLARE_DEFAULT_CONFIG(storage);
-static struct fs_mount_t fs_mnt = {
-	.type = FS_LITTLEFS,
-	.fs_data = &storage,
-	.storage_dev = (void *)FLASH_AREA_ID(storage),
-	.mnt_point = "/lfs",
-};
-#else
-#error No recognized file system
-#endif  /* file system */
-#endif  /* CONFIG_DISK_ACCESS_FLASH */
+#endif
 
-static void setup_disk(void)
+static struct fs_mount_t fs_mnt;
+
+static int setup_flash(void *storage_dev)
 {
+	int rc = 0;
 #if CONFIG_DISK_ACCESS_FLASH
-	struct fs_mount_t *mp = &fs_mnt;
-	unsigned int id = (uintptr_t)mp->storage_dev;
-	int rc;
+	unsigned int id = (uintptr_t)storage_dev;
 	const struct flash_area *pfa;
 
 	rc = flash_area_open(id, &pfa);
 	printk("Area %u at 0x%x on %s for %u bytes\n",
 	       id, (unsigned int)pfa->fa_off, pfa->fa_dev_name,
 	       (unsigned int)pfa->fa_size);
 
-	if (IS_ENABLED(CONFIG_APP_WIPE_STORAGE)) {
+	if (rc != 0 && IS_ENABLED(CONFIG_APP_WIPE_STORAGE)) {
 		printk("Erasing flash area ... ");
 		rc = flash_area_erase(pfa, 0, pfa->fa_size);
 		printk("%d\n", rc);
 	}
 
-	rc = fs_mount(&fs_mnt);
+	if (rc != 0) {
+		flash_area_close(pfa);
+	}
+#endif
+	return rc;
+}
+
+static int mount_fatfs(struct fs_mount_t *mnt)
+{
+	int rc = 0;
+#if CONFIG_FAT_FILESYSTEM_ELM
+	static FATFS fat_fs;
+
+	mnt->type = FS_FATFS;
+
+	if (IS_ENABLED(CONFIG_DISK_ACCESS_RAM)) {
+		mnt->mnt_point = "/RAM:";
+	} else {
+		mnt->mnt_point = "/NAND:";
+	}
+
+	mnt->storage_dev = DISK_STORAGE_DEV;
+	mnt->fs_data = &fat_fs;
+
+	rc = fs_mount(mnt);
+#endif
+	return rc;
+}
+
+static int mount_littlefs(struct fs_mount_t *mnt)
+{
+	int rc = 0;
+#if CONFIG_FILE_SYSTEM_LITTLEFS
+	if (!IS_ENABLED(CONFIG_DISK_ACCESS_FLASH)) {
+		return -ENOTSUP;
+	}
+
+	mnt->type = FS_LITTLEFS;
+	mnt->mnt_point = "/lfs";
+	mnt->storage_dev = DISK_STORAGE_DEV;
+	mnt->fs_data = &storage;
+
+	rc = fs_mount(mnt);
+#endif
+	return rc;
+}
+
+static void setup_disk(void)
+{
+	struct fs_mount_t *mp = &fs_mnt;
+	int rc;
+
+	if (!IS_ENABLED(CONFIG_DISK_ACCESS_RAM) &&
+	     IS_ENABLED(CONFIG_DISK_ACCESS_FLASH)) {
+		rc = setup_flash(mp->storage_dev);
+		if (rc != 0) {
```

Comment:
I think that most functions, this actually requires validation, will return negative value on error. Better check if `rc` is less then 0, just in case there would appear positive code that are non-error and do not require attention.

Reply:
I too would really like to see negative vs non-negative used to distinguish failure and success cases.  Zero is not always the only success case.